cgrok/grok.go

```diff
@@ -46,10 +49,23 @@ func New() *Grok {
 	if grok.g == nil {
 		return nil
 	}
-
+	grok.stringCache = make(map[uintptr]string)
 	return grok
 }
 
+/* Only look up group names, which live as long as the grok and are frequently reused. 
+   With other strings there's a risk of getting the wrong result if memory has been freed and reused. */
+func (grok *Grok) gostringn(str *C.char, len C.int) string {
```

Comment:
I think you've got it: the `uintptr` refers to the C string, which won't be freed for the life of the `grok` object. Right now it actually leaks, but since we allocate a fixed number of `grok`s we can get away with that. 

We can't use this caching strategy more generally because we can't guarantee that a pointer always refers to the same string. But in this case we know it won't be freed.